### PR TITLE
Fix WAL replay close hang (NFSE-5097) and add WAL data file upgrade support

### PR DESCRIPTION
**Pull request by nabeelmmd**
- Fix for wal replay close hang
- Add upgrade support to WAL data files

### CODE REVIEW

File: lib/c0/c0_kvmultiset.c [hunk 1]
```diff
@@ -977,6 +977,12 @@ c0kvms_putref(struct c0_kvmultiset *handle)
     ev(1);
 }
 
+void
+c0kvms_gen_init(u64 gen)
+{
+    atomic_set(&c0kvms_gen, gen);
+}
+
```
Comment (beckerg) [inline lib/c0/c0_kvmultiset.c:985]: uint64_t

**Review by beckerg — APPROVED** [review]

**Review by alexttx — APPROVED** [review]

File: lib/wal/wal_omf.h [hunk 2]
```diff
@@ -130,6 +161,27 @@ OMF_SETGET(struct wal_filehdr_omf, fh_endoff, 64);
...
+struct wal_rechdr_omf_v1 {
+    uint64_t rh_off;
+    uint32_t rh_flags;
+    uint32_t rh_cksum;
+    uint64_t rh_rid;
+    uint64_t rh_gen;
+    uint32_t rh_type;
+    uint32_t rh_len;
+    uint64_t rh_rsvd;
+} __attribute__((packed,aligned(sizeof(uint64_t))));
```
Comment (tristan957) [inline lib/wal/wal_omf.h:173]: Use attributes defined in compiler.h

Comment (beckerg) [reply lib/wal/wal_omf.h:173]: There is no such thing.  Probably we need an "HSE_OMF_ATTRS" or something that we'd put on all omf structures..

Comment (tristan957) [reply lib/wal/wal_omf.h:173]:
```c
#if defined(SUPPORTS_ATTR_PACKED) && defined(SUPPORTS_ATTR_ALIGNED)
#define HSE_OMF_ATTRS __attribute__((packed,aligned(sizeof(uint64_t))))
#else
#error "HSE requires the ability to pack and align on-media structures"
#endif
```

Something like this maybe? Not sure if the error is necessary, but sounds like without this there could be problems.

Comment (beckerg) [reply lib/wal/wal_omf.h:173]: Right.  There are some attributes that if they don't exist then hse likely either wont work or will run poorly.  Would rather know that they don't exist than to silently compile as if everything is ok.  Do we know of any compilers that don't support all the attributes we use?  xlc? icc?

Comment (tristan957) [reply lib/wal/wal_omf.h:173]: No clue...

Need to get HSE compiling on clang first though. Haven't tested it. Would also be nice to move off of gnu11 to c11 although I anticipate a lot of struggle there.

**Review by alexttx — APPROVED** [review]

File: lib/wal/wal_omf.h (hunk 2 above)
Comment (nabeelmmd) [reply lib/wal/wal_omf.h:173]: Thanks, we can do this cleanup in a different ticket..

**Pull request merged**